### PR TITLE
Make the register/unregisterCallback calls in StorageService non-blocking

### PR DESCRIPTION
**Pull request by jblebrun**
As part of this, the old `DeferredResult` helper class is removed,
replaced by the `suspendForCallback` family of helper methods. When
interacting with a service object that has a single callback.

Now that all of the service methods are oneway/async, and the on/off
methods of [ServiceStore] are suspend, we can now consistently use this
suspendForCallback-based pattern everywhere that we are calling into the
service.

### CODE REVIEW

File: java/arcs/android/storage/service/FutureRegistrationCallback.kt
Comment (jasonwyatt) [inline java/arcs/android/storage/service/FutureRegistrationCallback.kt:6]: KDoc needed

File: java/arcs/android/storage/service/FutureRegistrationCallback.kt
Comment (jasonwyatt) [inline java/arcs/android/storage/service/FutureRegistrationCallback.kt:13]:
```suggestion
    }
    
    override fun onFailure(exception: ByteArray?) {
```

File: java/arcs/android/storage/service/FutureResult.kt
Comment (jasonwyatt) [inline java/arcs/android/storage/service/FutureResult.kt:6]: KDoc needed

File: java/arcs/android/storage/service/FutureRegistrationCallback.kt
Comment (jasonwyatt) [reply java/arcs/android/storage/service/FutureRegistrationCallback.kt:6]: Also, I'm curious why you chose to use `Future` here, instead of `Deferred`?

File: java/arcs/android/storage/service/IRegistrationCallback.aidl
Comment (jasonwyatt) [inline java/arcs/android/storage/service/IRegistrationCallback.aidl:18]: Some documentation comments here explaining what the `byte[]` represents would be good. (e.g. that they're a protobuf serialization of a CrdtException).

File: java/arcs/sdk/android/storage/ServiceStore.kt
Comment (jasonwyatt) [inline java/arcs/sdk/android/storage/ServiceStore.kt:140]: Indentation needs some tweaking here

File: java/arcs/sdk/android/storage/ServiceStore.kt
Comment (jasonwyatt) [inline java/arcs/sdk/android/storage/ServiceStore.kt:145]: Ah, now I see why `Future` is preferred to `Deferred`... because we don't want to make `on` and `off` into suspend functions. 👍

File: java/arcs/android/storage/service/FutureRegistrationCallback.kt
Comment (jblebrun) [reply java/arcs/android/storage/service/FutureRegistrationCallback.kt:6]: It's used in non-suspending methods.

Comment (jblebrun) [reply java/arcs/android/storage/service/FutureRegistrationCallback.kt:6]: As you saw below :-)

File: java/arcs/android/storage/service/IRegistrationCallback.aidl
Comment (jblebrun) [reply java/arcs/android/storage/service/IRegistrationCallback.aidl:18]: will do doc/cleanup pass

File: java/arcs/sdk/android/storage/ServiceStore.kt
Comment (ianull) [inline java/arcs/sdk/android/storage/ServiceStore.kt:132]: What would happen if the service end died?

File: java/arcs/android/storage/service/BindingContext.kt
Comment (ianull) [inline java/arcs/android/storage/service/BindingContext.kt:140]: awaiting idle? typo?

File: java/arcs/sdk/android/storage/ServiceStore.kt
Comment (ianull) [inline java/arcs/sdk/android/storage/ServiceStore.kt:132]: What would happen if the service end died?

Feel there is a potential deadlock, i.e. thread pool size 2

timeline
```
thread #1: registerCallback -> scope.launch {} -> completion.get() // thread #1 is blocked
thread #2:               registerCallback -> scope.launch {} -> completion.get() // thread #2 is blocked
```
scope.launch {} for both threads can not be dispatched since both threads are blocked on waiting for onSuccess/onFailure in their scope.launch {} block. (deadlock)

File: java/arcs/android/storage/service/BindingContext.kt
Comment (jblebrun) [reply java/arcs/android/storage/service/BindingContext.kt:140]: good catch

File: java/arcs/sdk/android/storage/ServiceStore.kt
Comment (jblebrun) [reply java/arcs/sdk/android/storage/ServiceStore.kt:132]: Good catch on the death issue, I think I have it fixed now.

Comment (jblebrun) [reply java/arcs/sdk/android/storage/ServiceStore.kt:132]: This potential deadlock exists in the current version as well, right?

Comment (jblebrun) [reply java/arcs/sdk/android/storage/ServiceStore.kt:132]: The answer is yes! Today I've noticed that some tests started failing due to this exact deadlock (I'm not sure why the have started failing, and they are failing only in TAP, not on the GCB runs or on my machine). 

It seems to me that this is always going to be an issue: if we want `registerCallback` to be a synchronous/blocking function that returns a value, then we will be blocking a thread to wait for the result. 

Is there a solution to this other than ensuring that the threads that drive the StorageService are not the same as the threads that drive the ArcHosts?

We could make these methods (on/off/registerCallback/unregisterCallback). This might actually be a good idea, but will require some refactoring of StorageProxy, which currently indirectly triggers a call to `on` at construction time, so this method can not be suspending with the current structure.

Comment (ianull) [reply java/arcs/sdk/android/storage/ServiceStore.kt:132]: Sounds good to make on/off at ServiceStore suspending.

I remember there is also a todo to make the registerCallback one-way with callback.
This would also work with some refactoring i.e. make id getter Deferred<Int>, etc.

Another idea is a global unique id generator.
So we can still use scope.launch for service, in the meantime return an unique id back.
Clients get ids back asap and the service `associates` the assigned ids with callbacks simultaneously.

Comment (jblebrun) [reply java/arcs/sdk/android/storage/ServiceStore.kt:132]: This PR is the one making the registerCallback oneway w/ callback :-)

Comment (jblebrun) [reply java/arcs/sdk/android/storage/ServiceStore.kt:132]: I have some work in flight to convert `on` to a suspending method, then this PR can be converted to use Deferred instead of future.

Comment (ianull) [reply java/arcs/sdk/android/storage/ServiceStore.kt:132]: Cool! So glad we caught this and making it right :-)

**Comment by jblebrun** [comment]
Now that the `on` and `off` methods for stores are `suspending` (via some other recent PRs), I've updated this PR to take advantage of those, and now the register/on and unregister/off methods should be fully non-blocking.

**Comment by jblebrun** [comment]
CL to unblock copybara: cl/331017145 (can be landed independently)

File: java/arcs/android/storage/service/SuspendingServiceMethods.kt
Comment (jasonwyatt) [inline java/arcs/android/storage/service/SuspendingServiceMethods.kt:37]: Curious why you chose to do implementation delegation, rather than just extending the Stub?

**Review by jasonwyatt — APPROVED** [review]

File: java/arcs/android/storage/service/SuspendingServiceMethods.kt
Comment (jblebrun) [reply java/arcs/android/storage/service/SuspendingServiceMethods.kt:37]: Ah, good catch. Artifact of earlier iteration in which I had a common base class that I was extending. That's gone, so I can clean this up.

File: java/arcs/sdk/android/storage/ServiceStore.kt
Comment (jblebrun) [reply java/arcs/sdk/android/storage/ServiceStore.kt]: These are now both suspending.

**Pull request merged**
